kernel/aster-nix/src/net/iface/common.rs

```diff
@@ -147,56 +149,93 @@ impl IfaceCommon {
 
     pub(super) fn poll<D: Device + ?Sized>(&self, device: &mut D) {
         let mut interface = self.interface.lock_irq_disabled();
+        let mut sockets = self.sockets.lock_irq_disabled();
+
         let timestamp = get_network_timestamp();
-        let has_events = {
-            let mut sockets = self.sockets.lock_irq_disabled();
-            interface.poll(timestamp, device, &mut sockets)
-            // drop sockets here to avoid deadlock
-        };
-        if has_events {
-            self.bound_sockets.read().iter().for_each(|bound_socket| {
-                if let Some(bound_socket) = bound_socket.upgrade() {
-                    bound_socket.on_iface_events();
+        let (has_events, poll_at) = {
+            let mut has_events = false;
+            let mut poll_at;
+            loop {
+                has_events |= interface.poll(timestamp, device, &mut sockets);
+                poll_at = interface.poll_at(timestamp, &sockets);
+                let Some(instant) = poll_at else {
+                    break;
+                };
+                if instant > timestamp {
+                    break;
                 }
-            });
-        }
+            }
+            (has_events, poll_at)
+        };
+
+        // drop sockets here to avoid deadlock
+        drop(sockets);
+        drop(interface);
 
-        let sockets = self.sockets.lock_irq_disabled();
-        if let Some(instant) = interface.poll_at(timestamp, &sockets) {
-            let old_instant = self.next_poll_at_ms.load(Ordering::Acquire);
+        if let Some(instant) = poll_at {
+            let old_instant = self.next_poll_at_ms.load(Ordering::Relaxed);
             let new_instant = instant.total_millis() as u64;
             self.next_poll_at_ms.store(new_instant, Ordering::Relaxed);
 
-            if new_instant < old_instant {
+            if old_instant == 0 || new_instant < old_instant {
                 self.polling_wait_queue.wake_all();
             }
         } else {
             self.next_poll_at_ms.store(0, Ordering::Relaxed);
         }
+
+        if has_events {
+            self.bound_sockets.read().iter().for_each(|bound_socket| {
+                bound_socket.on_iface_events();
+            });
+
+            let closed_sockets = self
+                .closing_sockets
+                .lock()
```

Comment:
Maybe we should use `lock_irq_disabled` here? The whole `poll` function may be called in irq handler

Reply:
Well, we need to find a way to offload the `poll` method to the background poll thread or some softirq context in the future.

To clarify, I am not writing `lock_irq_disabled` here because we already have `self.bound_sockets.read()` before this PR. Should it be `self.bound_sockets.read_irq_disabled()`?

Reply:
> Well, we need to find a way to offload the poll method to the background poll thread or some softirq context in the future.

Yes, we need to the job in the future. At that time, the lock may not need to disable irq.

> To clarify, I am not writing lock_irq_disabled here because we already have self.bound_sockets.read() before this PR. Should it be self.bound_sockets.read_irq_disabled()?

I think it should be `read_irq_disabled`. This should be a mistake in the original code.